### PR TITLE
WIP play_by_name

### PR DESCRIPTION
**Pull request by ghost**
expand play_by_name functionality

it allows fun/useful things like

```
{
 "name"         : "Play TV From Beginning",
 "plugin"       : "plugin.video.meta",
 "live": [
  [
   {
    "link"      : "plugin://plugin.video.meta/tv/play_by_name/{program}/{language}",
    "action"    : "ACTIVATE"
   }
  ]
 ]
}
```

### CODE REVIEW

**Comment by metate** [comment]
To use different number of parameters use something like that (I think it's the best way, not sure though):

``` python
@plugin.route ("/live/<channel>/<program>/<language>", options={"program": "None", "language": "en"})
def live_play(channel, program, language):
     # code here: treat "None" as None
```

I can change xbmcswift/urls.py line 57 to use \* instead of +, this will allow passing empty strings in options (so you would be able to just use "if channel"). I will see if there are any drawbacks.

**Comment by ghost** [comment]
that would be nice, thank you.

File: plugin.video.meta/resources/lib/meta/navigation/tvshows.py [hunk 1]
```diff
@@ -100,10 +100,10 @@ def tv_search():
     """ Activate movie search """
     search(tv_search_term)
 
-@plugin.route('/tv/play_by_name/<name>/<season>/<episode>/<lang>')
+@plugin.route('/tv/play_by_name/<name>/<season>/<episode>/<lang>', options = {"season": "None", "episode": "None",
+                                                                              "lang": "en"})
 @plugin.route('/tv/play_by_name/<name>/<lang>', name="tv_play_by_name_without_episodes")
```
Comment (midraal) [inline plugin.video.meta/resources/lib/meta/navigation/tvshows.py]: left this one in as having the name and language does happen rather frequently, but the first route wont match this

File: plugin.video.meta/resources/lib/meta/navigation/tvshows.py [hunk 2]
```diff
@@ -100,8 +100,10 @@ def tv_search():
     """ Activate movie search """
     search(tv_search_term)
 
-@plugin.route('/tv/play_by_name/<name>/<season>/<episode>/<lang>')
-def tv_play_by_name(name, season, episode, lang = "en"):
+@plugin.route('/tv/play_by_name/<name>/<season>/<episode>/<lang>', options = {"season": "None", "episode": "None",
+                                                                              "lang": "en"})
+@plugin.route('/tv/play_by_name/<name>/<lang>', name="tv_play_by_name_without_episodes")
```
Comment (ghost) [inline plugin.video.meta/resources/lib/meta/navigation/tvshows.py]: left this one in as having the name and language does happen rather frequently, but the first route wont match this

**Comment by metate** [comment]
Seems good. Comments:

Instead of using selection+1 you should get the item by index and get the season/episode number from it (think about special season 0, missing episodes on tvdb, etc)

Put tv_play_by_name_without_episodes in a separate function that asks the user about season/episode and then invokes the original function (you can keep the same route). This is for clarity because if took me a minute to understand what is going on there.

I'll change "None" to "" after merging, and then make the tiny change to xbmcswift to handle it. My tests suggest that there is no regression.

I think I should test and push an update to the repo after this gets merged. Unless you plan other changes in this area that would invalidate the new endpoints. Please let me know.

**Comment by ghost** [comment]
how about something like this instead?

``` Python
@plugin.route('/tv/play_by_name/<name>/<season>/<episode>/<lang>', options = {"season": "None", "episode": "None",
                                                                              "lang": "en"})
@plugin.route('/tv/play_by_name/<name>/<lang>', name="tv_play_by_name_without_episodes")
def tv_play_by_name(name, season = None, episode = None, lang = "en"):
    """ Activate tv search """
    import_tvdb()

    search_results = tvdb.search(name, language= lang)

    if search_results == []:
        dialogs.ok(_("Show not found"), "{0} {1} in tvdb".format(_("no show information found for"), to_utf8(name)))
        return

    items = []
    for show in search_results:
        if "firstaired" in show:
            show["year"] = int(show['firstaired'].split("-")[0].strip())
        else:
            show["year"] = 0
        items.append(show)

    if len(items) > 1:
        selection = dialogs.select(_("Choose Show"), ["{0} ({1})".format(
            to_utf8(s["seriesname"]), s["year"]) for s in items])
    else:
        selection = 0
    if selection != -1:
        id = items[selection]["id"]
        if (not season or not episode) or (season == "None" or episode == "None"):
            season, episode = choose_season_and_episode(id)
        tv_play(id, season, episode, "default")

def choose_season_and_episode(tvdb_id):
    season = None
    episode = None

    while season is None or episode is None:
        show = tv_tvshow(tvdb_id)
        selection = dialogs.select(_("Choose season"), [item["label"] for item in show])
        if selection != -1:
            season = show[selection]["info"]["season"]
            season = int(season)
        else:
            return
        items = []
        episodes = tv_season(tvdb_id, season)
        for item in episodes:
            label = "S{0}E{1} - {2}".format(item["info"]["season"], item["info"]["episode"],
                                            to_utf8(item["info"]["title"]))
            if item["info"]["plot"] is not None:
                label += " - {0}".format(to_utf8(item["info"]["plot"]))
            items.append(label)
        selection = dialogs.select(_("Choose episode"), items)
        if selection != -1:
            episode = episodes[selection]["info"]["episode"]
            episode = int(episode)
        return season, episode
```

**Comment by metate** [comment]
That's fine as well. 

However, from what I can tell you either pass season&episose or you don't and let the user choose. So the two routes should be:
1. Required name, required season, required episode, optional language (default en)
2. Required name, optional language (default en)
Am I wrong about that?

Why the while in choose_season_and_episode?

**Comment by ghost** [comment]
no, those would be the 2 routes, but as both would need the same setup (getting tvdb_id, maybe pop up a selection for show) this seemed cleaner than copy-pasting it to a second function. maybe i could extract that part to another function, but not sure if it's worth it.

the while in choose_season_and_episode is so that if you get to the second select dialog and press back it goes back to the season selection instead of going all the way out

**Comment by ghost** [comment]
extracting the setup did make it a lot more readable

**Comment by metate** [comment]
Great, really clean now :)

**Comment by ghost** [comment]
[**@metate**](https://github.com/metate) commented on [Tuesday, April 12, 2016 12:16:49 PM](https://github.com/metate/meta4kodi/pull/35#issuecomment-208833933):

> I think I should test and push an update to the repo after this gets merged. Unless you plan other changes in this area that would invalidate the new endpoints. Please let me know.

i'm not planning anything invalidating the new endpoints, not realy sure what i'll work on next yet

**Pull request merged**